### PR TITLE
Add /grim history <player>

### PR DESCRIPTION
**Pull request by AbdisKiosk**
This PR introduces the possibility to display a player's violations by adding the /grim history <player> command. When triggered, the command will show something like this:
image

The PR also adds the "[log]" command for punishments.yml, which inserts the violation to the player's history

This change is extremely useful, as it makes it much easier to determine whether or not player is cheating.

### CODE REVIEW

**Comment by Anthony01M** [comment]
I highly recommend using hikaricp for this

File: src/main/java/ac/grim/grimac/manager/PunishmentManager.java [hunk 1]
```diff
@@ -150,6 +150,10 @@ public boolean handleAlert(GrimPlayer player, String verbose, Check check) {
                             if (command.command.equals("[webhook]")) {
                                 String vl = group.violations.values().stream().filter((e) -> e == check).count() + "";
                                 GrimAPI.INSTANCE.getDiscordManager().sendAlert(player, verbose, check.getDisplayName(), vl);
+                            } else if(command.command.equals("[log]")) {
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/manager/PunishmentManager.java]:
```suggestion
                            } else if (command.command.equals("[log]")) {
```

File: src/main/java/ac/grim/grimac/manager/log/Violation.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import java.util.Date;
+import java.util.UUID;
+
+@Data
+public class Violation {
+
+    private final UUID playerUUID;
+    private final String checkName;
+    private final String verbose;
+    private final String vl;
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/manager/log/Violation.java]: `double`?

File: src/main/java/ac/grim/grimac/manager/log/ViolationDatabaseManager.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+                                ")"
+                );
+        ) {
+            createTable.execute();
+
+            PreparedStatement createIndex = connection.prepareStatement(
+                    "CREATE INDEX IF NOT EXISTS idx_violations_uuid ON violations(uuid)"
+            );
+            createIndex.execute();
+        } catch(SQLException ex) {
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/manager/log/ViolationDatabaseManager.java]:
```suggestion
        } catch (SQLException ex) {
```

File: src/main/java/ac/grim/grimac/manager/log/ViolationDatabaseManager.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+                        "SELECT COUNT(*) FROM violations" +
+                                " WHERE uuid = ?"
+                )
+        ) {
+            fetchLogs.setString(1, player.toString());
+            ResultSet resultSet = fetchLogs.executeQuery();
+            if(resultSet.next()) {
+                return resultSet.getInt(1);
+            }
+        } catch(SQLException ex) {
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/manager/log/ViolationDatabaseManager.java]:
```suggestion
        } catch (SQLException ex) {
```

File: src/main/java/ac/grim/grimac/manager/log/ViolationDatabaseManager.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        try(
+                Connection connection = getConnection();
+                PreparedStatement fetchLogs = connection.prepareStatement(
+                        "SELECT COUNT(*) FROM violations" +
+                                " WHERE uuid = ?"
+                )
+        ) {
+            fetchLogs.setString(1, player.toString());
+            ResultSet resultSet = fetchLogs.executeQuery();
+            if(resultSet.next()) {
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/manager/log/ViolationDatabaseManager.java]:
```suggestion
            if (resultSet.next()) {
```

File: src/main/java/ac/grim/grimac/manager/log/ViolationDatabaseManager.java [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+            ResultSet resultSet = fetchLogs.executeQuery();
+            while(resultSet.next()) {
+                String checkName = resultSet.getString("check_name");
+                String verbose = resultSet.getString("verbose");
+                String vl = resultSet.getString("vl");
+                Date createdAt = new Date(resultSet.getLong("created_at"));
+
+                violations.add(new Violation(UUID.randomUUID(), checkName, verbose, vl, createdAt));
+            }
+        } catch(SQLException ex) {
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/manager/log/ViolationDatabaseManager.java]:
```suggestion
        } catch (SQLException ex) {
```

File: src/main/java/ac/grim/grimac/manager/log/ViolationDatabaseManager.java [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+                        "SELECT check_name, verbose, vl, created_at FROM violations" +
+                        " WHERE uuid = ? ORDER BY created_at DESC LIMIT ? OFFSET ?"
+                )
+        ) {
+            fetchLogs.setString(1, player.toString());
+            fetchLogs.setInt(2, limit);
+            fetchLogs.setInt(3, (page - 1) * limit);
+
+            ResultSet resultSet = fetchLogs.executeQuery();
+            while(resultSet.next()) {
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/manager/log/ViolationDatabaseManager.java]:
```suggestion
            while (resultSet.next()) {
```

File: src/main/java/ac/grim/grimac/manager/log/ViolationDatabaseManager.java [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+        } catch(SQLException ex) {
+            plugin.getLogger().log(Level.SEVERE, "Failed to fetch violations:", ex);
+        }
+
+        return violations;
+    }
+
+
+    protected Connection getConnection() throws SQLException {
+        if(openConnection == null || openConnection.isClosed()) {
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/manager/log/ViolationDatabaseManager.java]:
```suggestion
        if (openConnection == null || openConnection.isClosed()) {
```

File: src/main/java/ac/grim/grimac/manager/log/ViolationDatabaseManager.java [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+                    )
+            ) {
+                insertLog.setString(1, player.getUniqueId().toString());
+                insertLog.setString(2, verbose);
+                insertLog.setString(3, checkName);
+                insertLog.setString(4, violations);
+                insertLog.setLong(5, System.currentTimeMillis());
+
+                insertLog.executeUpdate();
+            } catch(SQLException ex) {
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/manager/log/ViolationDatabaseManager.java]:
```suggestion
            } catch (SQLException ex) {
```

File: src/main/java/ac/grim/grimac/manager/log/ViolationDatabaseManager.java [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+    public ViolationDatabaseManager(Plugin plugin) {
+        this.plugin = plugin;
+    }
+
+    @Override
+    public void start() {
+        try {
+            //Init sqlite
+            Class.forName("org.sqlite.JDBC");
+        } catch(ClassNotFoundException e) {
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/manager/log/ViolationDatabaseManager.java]:
```suggestion
        } catch (ClassNotFoundException e) {
```

File: src/main/java/ac/grim/grimac/manager/log/Violation.java [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+        return verbose;
+    }
+
+    public String getVl() {
+        return vl;
+    }
+
+    public Date getCreatedAt() {
+        return createdAt;
+    }
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/manager/log/Violation.java]: `@Getter`

File: src/main/java/ac/grim/grimac/commands/GrimHistory.java [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+
+        if (days > 0) result.append(days).append("d ");
+        if (hours > 0) result.append(hours).append("h ");
+        if (minutes > 0) result.append(minutes).append("m ");
+        if (seconds > 0) result.append(seconds).append("s");
+
+        return result.toString().trim();
+    }
+
+
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/commands/GrimHistory.java]:
```suggestion
```

File: src/main/java/ac/grim/grimac/commands/GrimHistory.java [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+            List<Violation> logs = violations.getViolations(target.getUniqueId(), notNullPage, 15);
+
+            int maxPages = (int) Math.ceil((float) logCount / 15);
+            sender.sendMessage(MessageUtil.format(header
+                    .replace("%player%", target.getName())
+                    .replace("%page%", String.valueOf(notNullPage))
+                    .replace("%maxPages%", String.valueOf(maxPages))
+            ));
+
+            for(int i = logs.size() - 1; i >= 0; i--) {
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/commands/GrimHistory.java]:
```suggestion
            for (int i = logs.size() - 1; i >= 0; i--) {
```

**Review by ManInMyVan — CHANGES_REQUESTED** [review]

File: src/main/java/ac/grim/grimac/commands/GrimHistory.java [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+public class GrimHistory extends BaseCommand {
+
+    @Subcommand("log|logs")
+    @CommandPermission("grim.history")
+    @CommandAlias("gh")
+    public void onLogs(CommandSender sender, OfflinePlayer target, @Optional Integer page) {
+        String header = GrimAPI.INSTANCE.getConfigManager().getConfig().getStringElse("grim-history-header",
+                "%prefix% §bShowing logs for §f%player% (§f%page%§b/§f%maxPages%§b)");
+        String logFormat = GrimAPI.INSTANCE.getConfigManager().getConfig().getStringElse("grim-history-log",
+                "%prefix% §bFailed §f%check% (x§c%vl%§f) §7%verbose% (§b%timeago% ago§7)");
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/commands/GrimHistory.java]: add these to the default messages configs

**Review by ManInMyVan — CHANGES_REQUESTED** [review]

File: src/main/java/ac/grim/grimac/commands/GrimHistory.java [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+
+@CommandAlias("grim|grimac")
+public class GrimHistory extends BaseCommand {
+
+    @Subcommand("log|logs")
+    @CommandPermission("grim.history")
+    @CommandAlias("gh")
+    public void onLogs(CommandSender sender, OfflinePlayer target, @Optional Integer page) {
+        String header = GrimAPI.INSTANCE.getConfigManager().getConfig().getStringElse("grim-history-header",
+                "%prefix% §bShowing logs for §f%player% (§f%page%§b/§f%maxPages%§b)");
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/commands/GrimHistory.java]:
```suggestion
                "%prefix% §bShowing logs for §f%player% (§f%page%§b/§f%maxPages%§f)");
```
I wonder if all white would look better for the page indicator

**Review by ManInMyVan — CHANGES_REQUESTED** [review]

File: src/main/java/ac/grim/grimac/manager/log/ViolationDatabaseManager.java [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+    private Connection openConnection;
+
+    public ViolationDatabaseManager(Plugin plugin) {
+        this.plugin = plugin;
+    }
+
+    @Override
+    public void start() {
+        try {
+            //Init sqlite
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/manager/log/ViolationDatabaseManager.java]:
```suggestion
            // Init sqlite
```

File: src/main/java/ac/grim/grimac/commands/GrimHistory.java [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+                "%prefix% §bFailed §f%check% (x§c%vl%§f) §7%verbose% (§b%timeago% ago§7)");
+
+        FoliaScheduler.getAsyncScheduler().runNow(GrimAPI.INSTANCE.getPlugin(), __ -> {
+            int notNullPage = page == null ? 1 : page;
+
+            ViolationDatabaseManager violations = GrimAPI.INSTANCE.getViolationDatabaseManager();
+            int logCount = violations.getLogCount(target.getUniqueId());
+            List<Violation> logs = violations.getViolations(target.getUniqueId(), notNullPage, 15);
+
+            int maxPages = (int) Math.ceil((float) logCount / 15);
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/commands/GrimHistory.java]: max log amount per page should be configurable

File: src/main/java/ac/grim/grimac/manager/log/ViolationDatabaseManager.java [hunk 18]
```diff
@@ -0,0 +1,139 @@
...
+            Class.forName("org.sqlite.JDBC");
+        } catch(ClassNotFoundException e) {
+            plugin.getLogger().log(Level.SEVERE, "Could not load SQLite driver", e);
+        }
+        try(
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/manager/log/ViolationDatabaseManager.java]:
```suggestion
        try (
```

File: src/main/java/ac/grim/grimac/manager/log/ViolationDatabaseManager.java [hunk 19]
```diff
@@ -0,0 +1,139 @@
...
+            );
+            createIndex.execute();
+        } catch(SQLException ex) {
+            plugin.getLogger().log(Level.SEVERE, "Failed to generate violations database:", ex);
+        }
+    }
+
+    public void logAlert(GrimPlayer player, String verbose, String checkName, String violations) {
+        FoliaScheduler.getAsyncScheduler().runNow(plugin, __ -> {
+            try(
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/manager/log/ViolationDatabaseManager.java]:
```suggestion
            try (
```

File: src/main/java/ac/grim/grimac/manager/log/ViolationDatabaseManager.java [hunk 20]
```diff
@@ -0,0 +1,139 @@
...
+
+                insertLog.executeUpdate();
+            } catch(SQLException ex) {
+                plugin.getLogger().log(Level.SEVERE, "Failed to insert violation:", ex);
+            }
+        });
+    }
+
+    public int getLogCount(UUID player) {
+        try(
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/manager/log/ViolationDatabaseManager.java]:
```suggestion
        try (
```

File: src/main/java/ac/grim/grimac/manager/log/ViolationDatabaseManager.java [hunk 21]
```diff
@@ -0,0 +1,139 @@
...
+                Connection connection = getConnection();
+                PreparedStatement fetchLogs = connection.prepareStatement(
+                        "SELECT COUNT(*) FROM violations" +
+                                " WHERE uuid = ?"
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/manager/log/ViolationDatabaseManager.java]:
```suggestion
                        "SELECT COUNT(*) FROM violations WHERE uuid = ?"
```

File: src/main/java/ac/grim/grimac/manager/log/ViolationDatabaseManager.java [hunk 22]
```diff
@@ -0,0 +1,139 @@
...
+            }
+        } catch(SQLException ex) {
+            plugin.getLogger().log(Level.SEVERE, "Failed to fetch number of violations:", ex);
+        }
+        return 0;
+    }
+
+    public List<Violation> getViolations(UUID player, int page, int limit) {
+        List<Violation> violations = new ArrayList<>();
+        try(
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/manager/log/ViolationDatabaseManager.java]:
```suggestion
        try (
```

**Review by ManInMyVan — CHANGES_REQUESTED** [review]

**Comment by AbdisKiosk** [comment]
> I highly recommend using hikaricp for this

Why?

File: src/main/java/ac/grim/grimac/manager/log/Violation.java (hunk 2 above)
Comment (AbdisKiosk) [reply src/main/java/ac/grim/grimac/manager/log/Violation.java]: For vl?

File: src/main/resources/config/en.yml [hunk 23]
```diff
@@ -186,4 +186,7 @@ max-ping-out-of-flying: 1000
 # This prevents high latency players from being able to use 1 firework boost with an elytra forever.
 max-ping-firework-boost: 1000
 
+# How many entries should be shown for each page with /grim history <player>
+grim-history-entries-per-page: 15
+
```
Comment (ManInMyVan) [inline src/main/resources/config/en.yml]: add this to all of the config files

File: src/main/resources/messages/en.yml [hunk 24]
```diff
@@ -50,4 +50,19 @@ help:
...
+# %page%
+# %maxPages%
+grim-history-header: "%prefix% §bShowing logs for §f%player% (§f%page%§b/§f%maxPages%§f)"
+# Valid placeholders:
+# %prefix%
+# %check%
+# %vl%
+# %verbose%
+# %timeago%
+grim-history-entry: "%prefix% §bFailed §f%check% (x§c%vl%§f) §7%verbose% (§b%timeago% ago§7)"
```
Comment (ManInMyVan) [inline src/main/resources/messages/en.yml]: add these to all of the messages files

**Review by ManInMyVan — CHANGES_REQUESTED** [review]

File: src/main/java/ac/grim/grimac/manager/log/Violation.java (hunk 2 above)
Comment (ManInMyVan) [reply src/main/java/ac/grim/grimac/manager/log/Violation.java]: yes

Comment (AbdisKiosk) [reply src/main/java/ac/grim/grimac/manager/log/Violation.java]: It should be an `int` if anything. I just copied the code from the DiscordManager. I'll change it tomorrow

**Comment by Dg32z** [comment]
> > I highly recommend using hikaricp for this
> 
> Why?

I think this may enable querying on any server that uses Grim

**Comment by Dg32z** [comment]
> > > I highly recommend using hikaricp for this
> > 
> > 
> > Why?
> 
> I think this may enable querying on any server that uses Grim

If possible, I strongly recommend synchronizing Violation across different servers (:

**Comment by AbdisKiosk** [comment]
> > > > I highly recommend using hikaricp for this
> 
> > > 
> 
> > > 
> 
> > > Why?
> 
> > 
> 
> > I think this may enable querying on any server that uses Grim
> 
> 
> 
> If possible, I strongly recommend synchronizing Violation across different servers (:

Using hikaricp doesn't automatically make querying across servers possible. However, I could implement MySQL support pretty easily, but that would require users to have their own MySQL database.

File: src/main/resources/messages/en.yml [hunk 25]
```diff
@@ -50,4 +50,19 @@ help:
   - "/grim spectate <player> &f- &7Spectate a player"
   - "/grim verbose &f- &7Shows every flag to you, without buffers"
   - "/grim log [0-255] &f- &7Uploads a debug log for prediction flags"
+  - "/grim history <player> [page] §f- §7Shows previous alerts the player"
```
Comment (ManInMyVan) [inline src/main/resources/messages/en.yml]: add this to all of the messages files

**Review by ManInMyVan — CHANGES_REQUESTED** [review]

File: src/main/java/ac/grim/grimac/manager/violationdatabase/SQLiteViolationDatabase.java [hunk 26]
```diff
@@ -0,0 +1,134 @@
...
+    public void connect() {
+        try {
+            // Init sqlite
+            Class.forName("org.sqlite.JDBC");
+        } catch (ClassNotFoundException e) {
+            plugin.getLogger().log(Level.SEVERE, "Could not load SQLite driver", e);
+        }
+        try (
+                Connection connection = getConnection();
+        ) {
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/manager/violationdatabase/SQLiteViolationDatabase.java]:
```suggestion
        try (Connection connection = getConnection()) {
```

**Review by ManInMyVan — CHANGES_REQUESTED** [review]

File: src/main/java/ac/grim/grimac/commands/GrimHistory.java [hunk 27]
```diff
@@ -0,0 +1,86 @@
...
+
+    @Subcommand("history|hist")
+    @CommandPermission("grim.history")
+    @CommandAlias("gh")
+    @CommandCompletion("@players")
+    public void onLogs(CommandSender sender, String target, @Optional Integer page) {
+        int entriesPerPage = GrimAPI.INSTANCE.getConfigManager().getConfig().getIntElse("history.entries-per-page", 15);
+
+        String header = GrimAPI.INSTANCE.getConfigManager().getConfig().getStringElse("grim-history-header",
+                "%prefix% §bShowing logs for §f%player% (§f%page%§b/§f%maxPages%§f)");
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/commands/GrimHistory.java]:
```suggestion
                "%prefix% &bShowing logs for &f%player% (&f%page%&b/&f%maxPages%&f)");
```

File: src/main/java/ac/grim/grimac/commands/GrimHistory.java [hunk 28]
```diff
@@ -0,0 +1,86 @@
...
+        String logFormat = GrimAPI.INSTANCE.getConfigManager().getConfig().getStringElse("grim-history-entry",
+                "%prefix% §bFailed §f%check% (x§c%vl%§f) §7%verbose% (§b%timeago% ago§7)");
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/commands/GrimHistory.java]:
```suggestion
                "%prefix% &bFailed &f%check% (x&c%vl%&f) &7%verbose% (&b%timeago% ago&7)");
```

**Review by ManInMyVan — CHANGES_REQUESTED** [review]

File: gradlew
Comment (ManInMyVan) [inline gradlew:1]: Why was this changed?

**Review by ManInMyVan — CHANGES_REQUESTED** [review]

File: gradlew
Comment (AbdisKiosk) [reply gradlew:1]: don't know. i've reverted it now. sorry

**Comment by Anthony01M** [comment]
did a quick review, you need to add `server` to the schema to be able to handle cross-servers. 
`[%server%] %user% failed %check_name% [%time%]` (simple example I didn't check your placeholders)

**Comment by AbdisKiosk** [comment]
> did a quick review, you need to add `server` to the schema to be able to handle cross-servers. `[%server%] %user% failed %check_name% [%time%]` (simple example I didn't check your placeholders)

great suggestion, i've added it now

**Comment by iam-best** [comment]
Yes, please!

**Comment by ManInMyVan** [comment]
@AbdisKiosk please open a new pull request for this so it can this can be reviewed and merged

**Comment by iam-best** [comment]
Wait is it merged now and available or not?

**Comment by ManInMyVan** [comment]
> Wait is it merged now and available or not?

no, axionize accidentally merged this pr instead of the one in his fork, and reverted the merge

**Comment by iam-best** [comment]
:/

**Comment by Axionize** [comment]
@iam-best you can get my fork with this feature along with a slew of other improvements that I hope will (eventually) all make their way into Grim here: https://modrinth.com/plugin/lightning-grim-anticheat

**Comment by iam-best** [comment]
@Axionize thank you so much, just what I needed a simple additional plugin that adds this Feature! :)

**Comment by iam-best** [comment]
@Axionize do I need to remove grimac or have both?

**Comment by Axionize** [comment]
LightningGrim is a fork of Grim. Its is not an extension.

Remove upsteram GrimAC from your plugins folder and use only LightningGrim.

**Comment by iam-best** [comment]
@Axionize ok thanks!

**Comment by iam-best** [comment]
@Axionize can you maybe add a feature in the future that connects to the DiscordSRV Plugin and everytime someone hacks it pings a user or role if you haven't added it already? Thanks also maybe change that you can also see the logs from all Players at the same time!

**Comment by Anthony01M** [comment]
Incorrect place to ask for, discordsrv is also heavy to have on your
server, we can easily make it so that the webhook system is much more
advanced than it currently is and takes discord limitations into
consideration and also does what you need.

**Comment by iam-best** [comment]
@Anthony01M if you have alternative plugins you can DM them me of reply here then :)

**Comment by Anthony01M** [comment]
None, simply use the webhook system already built-in, that's good enough,
instead of adding more load on your server.

**Comment by iam-best** [comment]
@Anthony01M yeah I use it the plugin for other stuff too but how can I do that... Is there a tutorial? If it does the same thing as the Discord Plugin I will remove it but first need an alternative.

**Pull request merged**